### PR TITLE
L1-282: Exponential inflation

### PR DESCRIPTION
**Pull request by lesniak43**
# Description

Introducing exponential inflation.

## Type of change

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

### CODE REVIEW

File: pallets/aleph/src/lib.rs [hunk 1]
```diff
@@ -68,6 +69,27 @@ pub mod pallet {
...
+        const MILLISECS_PER_YEAR: u64 = 1000 * 3600 * 24 * 36525 / 100;
+        MILLISECS_PER_YEAR
+    }
+
+    #[pallet::storage]
+    pub type AzeroCap<T: Config> = StorageValue<_, Balance, ValueQuery, DefaultAzeroCap>;
+
+    #[pallet::storage]
+    pub type ExponentialInflationHorizon<T: Config> =
+        StorageValue<_, u64, ValueQuery, DefaultExponentialInflationHorizon>;
```
Comment (kostekIV) [inline pallets/aleph/src/lib.rs:91]: call to set these storages is missing

**Review by DamianStraszak — APPROVED** [review]
Apart from the storage setters this is good to go.

I'm assuming that once we know which/if additional API is necessary to display necessary info in the frontends, then it will be added in a separate PR.

File: bin/runtime/src/lib.rs [hunk 2]
```diff
@@ -441,11 +441,35 @@ parameter_types! {
...
 }
 
-pub struct UniformEraPayout;
+pub struct ExponentialEraPayout;
 
-impl pallet_staking::EraPayout<Balance> for UniformEraPayout {
-    fn era_payout(_: Balance, _: Balance, era_duration_millis: u64) -> (Balance, Balance) {
-        primitives::staking::era_payout(era_duration_millis)
+/// Calculates 1 - exp(-x) for small positive x
+fn exp_helper(x: Perbill) -> Perbill {
```
[the remark this answers is not in the document]
Comment (lesniak43) [reply bin/runtime/src/lib.rs:447]: rewrote the function to make the computation more explicit and ensure that `1 - exp(-x) < x`

**Review by kostekIV — APPROVED** [review]
looks good.

File: bin/runtime/src/lib.rs [hunk 3]
```diff
@@ -441,11 +441,35 @@ parameter_types! {
...
+    let x2 = x * x;
+    let x3 = x2 * x;
+    let x4 = x2 * x2;
+    let x5 = x4 * x;
+    (x - x2 / 2 + x3 / 6 - x4 / 24 + x5 / 120).min(x)
```
Comment (DamianStraszak) [inline bin/runtime/src/lib.rs:452]: Hmm, what happens in case of overflows in multiplication? Are we sure we never panic?

**Pull request merged**